File: 백준/Gold/23352. 방탈출/방탈출.py
import sys
from collections import deque

N, M = map(int, sys.stdin.readline().split())

graph = []

for i in range(N):
    graph.append(list(map(int, sys.stdin.readline().split())))

maxcnt = 0
answer = 0

def BFS(startx, starty):
    global maxcnt
    global answer

    visited = [[False] * M for _ in range(N)]
    q = deque()
    q.append([startx, starty, 0])
    visited[startx][starty] = True

    nx = [-1, 1, 0, 0]
    ny = [0, 0, -1, 1]

    while q:
        mylist = q.popleft()
        x = mylist[0]
        y = mylist[1]
        cnt = mylist[2]
        doyougo = False

        for i in range(4):
            dx = x + nx[i]
            dy = y + ny[i]

            if 0 <= dx < N and 0 <= dy < M and graph[dx][dy] > 0 and not visited[dx][dy]:
                doyougo = True
                visited[dx][dy] = True
                q.append([dx, dy, cnt + 1])

        if not doyougo:
            # 더 갈 곳이 없다면
            if maxcnt < cnt:
                maxcnt = cnt
                answer = graph[startx][starty] + graph[x][y]

            elif maxcnt == cnt:
                
                answer = max(answer, graph[startx][starty] + graph[x][y])

for i in range(N):
    for j in range(M):
        if graph[i][j] > 0:
            BFS(i, j)

print(answer)
# 출발지 0이 아니어야.

File: 백준/Gold/23352. 방탈출/방탈출.java
import java.io.*;
import java.util.*;

public class Main {
    static int N, M;
    static int[][] graph;
    static int maxCnt = 0;
    static int answer = 0;

    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] firstLine = br.readLine().split(" ");
        N = Integer.parseInt(firstLine[0]);
        M = Integer.parseInt(firstLine[1]);

        graph = new int[N][M];

        for (int i = 0; i < N; i++) {
            String[] line = br.readLine().split(" ");
            for (int j = 0; j < M; j++) {
                graph[i][j] = Integer.parseInt(line[j]);
            }
        }

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (graph[i][j] > 0) {
                    bfs(i, j);
                }
            }
        }

        System.out.println(answer);
    }

    static void bfs(int startX, int startY) {
        boolean[][] visited = new boolean[N][M];
        Queue<int[]> q = new LinkedList<>();
        q.add(new int[]{startX, startY, 0}); // x, y, cnt
        visited[startX][startY] = true;

        while (!q.isEmpty()) {
            int[] cur = q.poll();
            int x = cur[0];
            int y = cur[1];
            int cnt = cur[2];
            boolean hasNext = false;

            for (int dir = 0; dir < 4; dir++) {
                int nx = x + dx[dir];
                int ny = y + dy[dir];

                if (nx >= 0 && ny >= 0 && nx < N && ny < M) {
                    if (!visited[nx][ny] && graph[nx][ny] > 0) {
                        hasNext = true;
                        visited[nx][ny] = true;
                        q.add(new int[]{nx, ny, cnt + 1});
                    }
                }
            }

            if (!hasNext) {
                if (cnt > maxCnt) {
                    maxCnt = cnt;
                    answer = graph[startX][startY] + graph[x][y];
                } else if (cnt == maxCnt) {
                    answer = Math.max(answer, graph[startX][startY] + graph[x][y]);
                }
            }
        }
    }
}
